module/VuFind/src/VuFind/ILS/Logic/OnlinePaymentTrait.php

```diff
@@ -0,0 +1,144 @@
+ * @package  ILS
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+trait OnlinePaymentTrait
+{
+    /**
+     * Return details on fees payable online.
+     *
+     * @param array  $patron          Patron
+     * @param array  $fines           Patron's fines
+     * @param ?array $selectedFineIds Selected fines
+     *
+     * @throws ILSException
+     * @return array Associative array of payment details
+     *
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function getOnlinePaymentDetails(array $patron, array $fines, ?array $selectedFineIds): array
+    {
+        $amount = 0;
+        $payableFines = [];
+        foreach ($fines as $fine) {
+            // Nothing can be paid if there are blocking fines:
+            if ($this->fineBlocksPayment($fine)) {
+                return [
+                    'payable' => false,
+                    'amount' => 0,
+                    'fines' => [],
+                    'reason' => 'Payment::fines_contain_nonpayable_fees',
+                ];
+            }
+            if (
+                null !== $selectedFineIds
+                && !in_array($fine['fine_id'] ?? '', $selectedFineIds)
+            ) {
+                continue;
+            }
+            if ($fine['payable_online'] ?? false) {
+                $amount += $fine['balance'];
+                $payableFines[] = $fine;
+            }
+        }
+        return [
+            'payable' => $amount > 0,
+            'amount' => $amount,
+            'fines' => $payableFines,
+        ];
+    }
+
+    /**
+     * Check if a fine is payable.
+     *
+     * @param array $fine Fine
+     *
+     * @return bool
+     */
+    protected function fineIsPayable(array $fine): bool
+    {
+        if ($fine['balance'] <= 0) {
+            return false;
+        }
+        $paymentConfig = $this->config['OnlinePayment'] ?? [];
+        if (in_array($fine['fine'] ?? '', $paymentConfig['nonPayableTypes'] ?? [])) {
+            return false;
+        }
+        foreach ((array)($paymentConfig['nonPayableDescriptions'] ?? []) as $pattern) {
+            if (preg_match($pattern, $fine['description'] ?? '')) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Check if a fine should completely block payment.
+     *
+     * @param array $fine Fine
+     *
+     * @return bool
+     */
+    protected function fineBlocksPayment(array $fine): bool
+    {
+        $paymentConfig = $this->config['OnlinePayment'] ?? [];
+        if (in_array($fine['fine'] ?? '', $paymentConfig['blockingNonPayableTypes'] ?? [])) {
+            return true;
+        }
+        foreach ((array)($paymentConfig['blockingNonPayableDescriptions'] ?? []) as $pattern) {
+            if (str_starts_with($pattern, '/') && preg_match('{/\w?$}', $pattern)) {
```

Comment:
If we're using a regex anyway, would it be just as simple to:

```suggestion
            if (preg_match('{^/.*/\w?$}', $pattern)) {
```

Reply:
Sure, but if the pattern doesn't start with a slash, we can avoid compiling a regex. I know the perf difference is negligible, but I still like to avoid running a regex if it's not required.

Reply:
Fair enough!